### ISSUE TITLE
Permissions issues? Missing folders?

### ISSUE THREAD

**Issue by Saeven** [umpirsky/Twig-Gettext-Extractor#12]
Any of you developers using this on OSX?  I'm running into an issue where the call to your library:

```
<project>/vendor/bin/twig-gettext-extractor --sort-output --force-po -o "/var/folders/lw/0f57k9dj617_ntr760qw2lcr0000fgp/T/poedit1nmsrR/3extracted.pot" --from-code=UTF-8  -k_ -ktranslate -ksetLabel -ksetValue -ksetLegend -k_refresh -L PHP --files  "vendor/mylibrary/core/view/ldp/forgot/email.twig" "vendor/mylibrary/core/view/ldp/forgot/index.twig" "vendor/mylibrary/core/view/ldp/login/index.twig" "vendor/mylibrary/core/view/ldp/main/index.twig" "vendor/mylibrary/core/view/ldp/register/index.twig"
```

Fails with this error message:

```
Fatal error: Uncaught exception 'RuntimeException' with message 'Gettext command "xgettext --sort-output --force-po -o /var/folders/lw/0f57k9dj617_ntr760qw2lcr0000fgp/T/poedit1nmsrR/3extracted.pot --from-code=UTF-8 -k_ -ktranslate -ksetLabel -ksetValue -ksetLegend -k_refresh -L PHP /tmp/cache/5279bc643f96f/8c/e1/c60ddc5ef854138efa25631b4b6d.php /tmp/cache/5279bc643f96f/d3/37/bf84f1be48ed8fd950ff618fba22.php /tmp/cache/5279bc643f96f/a7/e7/56cec92524ceab7ab465d688f936.php /tmp/cache/5279bc643f96f/73/df/2a5e4848ca69cb43dc05818ce5b0.php /tmp/cache/5279bc643f96f/82/0b/167d3083d340e123ea9742944113.php" failed with error code 1 and output: ' in <project>/vendor/umpirsky/twig-gettext-extractor/Twig/Gettext/Extractor.php:90
```

So attempting to run that command directly leads to:

```
xgettext: cannot create output file "/var/folders/lw/0f57k9dj617_ntr760qw2lcr0000fgp/T/poedit1nmsrR/3extracted.pot": No such file or directory
```

All PHP these PHP scripts are owned by Me:admin on my filesystem.  Seen this before?

**Comment by barmarko27** [comment]
I'm on OSX Yosemite and the issue of missing folder under "/var" still continue. There are many OSX users that have same problem? Thank so much!

**Comment by umpirsky** [comment]
@barmarko27 Have you tried https://github.com/umpirsky/Twig-Gettext-Extractor/issues/12#issuecomment-44317644?

**Comment by barmarko27** [comment]
@umpirsky Yes, but don't work! It seems that Poedit can't delete the temp file. 

Poedit parser configuration: 
![schermata 2015-05-26 alle 12 28 41](https://cloud.githubusercontent.com/assets/6019223/7810567/5a737f3e-03a3-11e5-96f4-1b6db33395f1.png)

Poedit Command Line Execution:
![schermata 2015-05-26 alle 12 31 23](https://cloud.githubusercontent.com/assets/6019223/7810582/8a1ffd98-03a3-11e5-9d7a-3b5f078301cf.png)

Poedit Error:
![schermata 2015-05-26 alle 12 31 07](https://cloud.githubusercontent.com/assets/6019223/7810602/a2d1f29c-03a3-11e5-852d-e99de4cacda7.png)

Thanks for your support!

**Comment by umpirsky** [comment]
@arekkas Any idea why it does not work for @barmarko27? Thanks.

**Comment by arekkas** [comment]
With the latest POedit version, you don't need `WXTRACE=poedit,poedit.tmp,poedit.execute /Applications/Poedit.app/Contents/MacOS/Poedit --verbose --keep-temp-files` any more. Just start PoEdit regularly.

You'll also need to make sure that the path to `twig-gettext-extractor` is correct by running it in the console.

**Comment by barmarko27** [comment]
@arekkas Yes the path to twig-gettext-extractor is correct. It seems that Poedit can't delete temporary folder after execution of Twig Parser. Thanks for helping me!

**Comment by umpirsky** [comment]
Is it a permission problem?

**Comment by arekkas** [comment]
@barmarko27 did you start POEdit without `--keep-temp-files`?

**Comment by saadhre** [comment]
Platform: MacOS Yosemite 10.10.4 (14E46)
Poeditor:  1.8.3 (3998) with xgettext 0.19.4. 

When I starting Poeditor without parameters and trying to update translation from source, poeditor throws described error. With --keep-temp-files or with WXTRACE=poedit,poedit.tmp,poedit.execute exception is different:

Can't remove directory '/var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditQ3Mdg2' (error 66: Directory not empty)
Can't run application: PATH_TO_PROJECT/vendor/bin/twig-gettext-extractor --sort-output --force-po -o "/var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditQ3Mdg2/14extracted.pot" --from-code=UTF-8  -ktranslate -L PHP --files  "Application/view/application/index/index.twig"

Any idea?

**Comment by arekkas** [comment]
Can't run application: **PATH_TO_PROJECT**/vendor/bin/twig-gettext-extractor --sort-output --force-po -o "/var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditQ3Mdg2/14extracted.pot" --from-code=UTF-8 -ktranslate -L PHP --files "Application/view/application/index/index.twig"

Doesn't look like a real path to me ;)

**Comment by saadhre** [comment]
Yes, real path has sensitive client data in names. But trust me - path is correct.

**Comment by arekkas** [comment]
I see :)
Does verbose give useful info?

**Comment by saadhre** [comment]
In trace everything looks fine... This trace data before exception:

Trace: (poedit.tmp) new temp file /var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditmG51sW/14extracted.pot
Trace: (poedit.execute) /Applications/Poedit.app/Contents/PlugIns/GettextTools.bundle/Contents/MacOS/bin//Users/urfin/Public/www/******/******/vendor/bin/twig-gettext-extractor doesn't exist, falling back to /Users/urfin/Public/www/******/******/vendor/bin/twig-gettext-extractor
Trace: (poedit.execute) executing: /Users/urfin/Public/www/******/******/vendor/bin/twig-gettext-extractor --sort-output --force-po -o "/var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditmG51sW/14extracted.pot" --from-code=UTF-8  -ktranslate -L PHP --files  "Application/view/application/index/index.twig"
Trace: (poedit.tmp) keeping temp files in /var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditmG51sW
Trace: (poedit.tmp) removing temp dir /var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditmG51sW

**Comment by saadhre** [comment]
Just checked content of /var/folders/7f/hv40gb8x4cgg0v7wsbkm4djr0000gn/T/poeditmG51sW/ directory and 14extracted.pot not exists in there.

**Comment by ternes3** [comment]
same problem here.
start with --keep-temp-files ends in error 66#
without --keep-temp-files ends in 

Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/0extracted.pot
Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/1extracted.pot
Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/2extracted.pot
Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/3extracted.pot
Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/4extracted.pot
Trace: (poedit.tmp) removing temp file /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08/5merged.pot
Trace: (poedit.tmp) removing temp dir /var/folders/cz/3_w92y755tg8ksbss737427m0000gn/T/poeditwRmF08

**Comment by jeremyzahner** [comment]
Having the same problem:

````
Trace: (poedit.tmp) new temp file /var/folders/nw/qt8p5kc174dfrw4r5pf3c67h0000gn/T/poeditZoZrEQ/0extracted.pot
Trace: (poedit.execute) /Applications/Poedit.app/Contents/PlugIns/GettextTools.bundle/Contents/MacOS/bin//Users/jzahner/Sites/joshmartin-website/vendor/bin/twig-gettext-extractor doesn't exist, falling back to /Users/jzahner/Sites/joshmartin-website/vendor/bin/twig-gettext-extractor
Trace: (poedit.execute) executing: /Users/jzahner/Sites/joshmartin-website/vendor/bin/twig-gettext-extractor --sort-output --force-po -o "/var/folders/nw/qt8p5kc174dfrw4r5pf3c67h0000gn/T/poeditZoZrEQ/0extracted.pot" --from-code=UTF-8  -L PHP --files  "_angebot.twig" "_index.twig" "_kontakt.twig" "_team.twig" "_technologie.twig" "_werte.twig" "baselayout.twig"
Trace: (poedit.tmp) keeping temp files in /var/folders/nw/qt8p5kc174dfrw4r5pf3c67h0000gn/T/poeditZoZrEQ
Trace: (poedit.tmp) removing temp dir /var/folders/nw/qt8p5kc174dfrw4r5pf3c67h0000gn/T/poeditZoZrEQ
````

Any Updates on this?

**Comment by umpirsky** [comment]
@jeremyzahner MacOS?

**Comment by jeremyzahner** [comment]
Yes! Sorry for leaving out that detail ^^

**Comment by umpirsky** [comment]
@jeremyzahner The problem is that I am a Linux user myself and project is tested on OSX by other users. Have you tried @arekkas suggestion from https://github.com/umpirsky/Twig-Gettext-Extractor/issues/12#issuecomment-44317644 ?

**Comment by arekkas** [comment]
I can confirm this issue with recent POEdit versions. My best guess is to rollback your POEdit version.

**Comment by umpirsky** [comment]
Thanks for the update @arekkas. Let's try to ping Mr. @vslavik the author of Poedit, maybe he can give us some inside information about Poedit and how it handles files.

**Comment by vslavik** [comment]
> Let's try to ping Mr. @vslavik the author of Poedit, maybe he can give us some inside information about Poedit and how it handles files.

Sorry, it’s not really clear what you’re asking about — this issue is a bit of a mess to read through and I can’t quite make sense of all it. If you have a reproducible test case (one that doesn’t take half an eternity to setup), by all means, send it to me to have a look.

Without that, just a few assorted observations:

1. The original issue doesn’t seem related to Poedit: it’s clearly some extractor failure. The subsequent complain about manual `xgettext` command failing is bogus: of course it fails if you didn’t create the output directory (as Poedit would before invoking the command).

2. The advice to always use `--keep-temp-files` doesn’t make sense to me — that’s for debugging only!

3. The errors about failing to delete the temp directory are *caused* by the use of `--keep-temp-files` (due to a minor bug where Poedit would still try to delete the dir despite the flag — I’ll fix that right away).

4. If some `*extracted.pot` file is not created and not kept with `--keep-temp-files` (this is the only valid use for it: debugging if some files were created or not), then the only explanation is that the extractor tool that was supposed to create it didn't. Notice how the builtin extractors use the `--force-po` to always create an output file, even if empty. If `twig-gettext-extractor` doesn’t create an empty-but-valid POT file (i.e. with header, but no other msgids) when it doesn’t encounter any translatable strings, that would certainly cause this.

5. The use of relative paths to extractor executable was never supported and if it worked at some point, it was by pure accident. It’s not really clear to me why would you want to do that (instead of having a copy of `twig-gettext-extractor` somewhere in PATH to use with all your projects), but I assume I must be missing something, and would welcome education (probably at help@poedit.net as it would be offtopic here).

(BTW, I’m reworking extractors completely for the next major version (don’t worry, there will be “legacy” compatibility) and am soliciting any extractors-related feedback from, you know, real users of custom extractors — feel free to ping me at the above email address).

**Comment by arekkas** [comment]
> The errors about failing to delete the temp directory are caused by the use of --keep-temp-files (due to a minor bug where Poedit would still try to delete the dir despite the flag — I’ll fix that right away).

This will fix our issue 100%. It would be great if you could implement that!

**Comment by vslavik** [comment]
> This will fix our issue 100%. It would be great if you could implement that!

I feel like I might have not been as clear as I thought I was, so let me repeat myself in the interest of being absolutely clear:

**DO NOT EVER USE `--keep-temp-files`!**

It doesn’t solve anything and is not meant to be used by end users! If it *does* seem to help you with something, please show me a reproducible test case of *what* it solves, using the latest Poedit version, and I’ll fix *that*, because that would uncover some serious issue — but to be frank, I simply don’t see how could anything of the sorts be possible.

**Comment by arekkas** [comment]
Well I believe that the issue is as follows:

twig-gettext-extractor converts *.twig (no native POEdit support) files to *.php files. These get stored in a temp dir and then iterated by xgettext. Somewhere inbetween, these files get deleted which is why xgettext is saying that they don't exist. This issue was "resolved" by telling POEdit to keep the files.

That's at least what I think is happening.

**Comment by vslavik** [comment]
I see two problems with that theory:

1. Why do you think twig-gettext-extractor’s private temp files are stored in Poedit’s private temporary directory? Poedit creates a temp (sub)directory of its own, doesn’t change any environment variables and doesn’t change CWD to the tempdir either. Unless the extractor goes to extra pain to set the cache directory to Poedit’s temp one (i.e. the location of the *output file*, not the input twig files), I’m not sure how could that happen? If it does happen, it would indeed be problematic for Poedit (and I’ll address that too, because there’s no reason why Poedit shouldn’t survive some leftover temp files), but is it happening?

2. Looking at the code, the extractor generates PHP files, runs xgettext and only exits after that. Poedit doesn’t, and cannot, delete the temp folder before `twig-gettext-extractor` returns. There’s no way for it to delete files from under the extractor’s hands, even if they were located in Poedit’s temp directory.

I’m more and more confused by the minute :-/

**Comment by arekkas** [comment]
I'm sorry but I'm not sure if I can help you further. I ran into this issue some years ago and found the `--keep-temp-files` fix in some blog. It worked back then but at some point I upgraded POEdit and it stopped working. It has not been working ever since. I also think that this issue is Mac only. Maybe Mac manages its temp directory in a different way than Linux. No idea.

**Comment by umpirsky** [comment]
Thanks for your response and willing to help @vslavik, I really appreciate it.

One thing is sure. I am 100% sure cache directories of [Twig](https://github.com/umpirsky/Twig-Gettext-Extractor/blob/master/twig-gettext-extractor#L26) and Poedit are not shared. No doubt about that.

I never encouraged people to use `--keep-temp-files` it was never in readme. I see they use it for debug purposes.

I agree this issue is too verbose and hard to read, so for the sake of sanity let's just wait for some Mac user to come up with proper and easy to reproduce test case.

Now we have @vslavik who is willing to take a look and I am very thankful. :+1:

**Comment by vslavik** [comment]
> One thing is sure. I am 100% sure cache directories of Twig and Poedit are not shared.

Still a mystery, then. Let’s wait for more details as you say, then.

> I never encouraged people to use --keep-temp-files

Of course! I just wanted to clarify this for readers of this issue.

**Comment by Saeven** [comment]
Keep temp files did indeed fix things for me once upon a time:
https://groups.google.com/d/msg/poedit/t1HzAIZxKUc/SRjp2bAwYm0J

It was in a debug command that @vslavik had recommended during a test run.

**Comment by vslavik** [comment]
> Keep temp files did indeed fix things for me once upon a time:
> https://groups.google.com/d/msg/poedit/t1HzAIZxKUc/SRjp2bAwYm0J

According to the archive it was the act of *running Poedit from command line* that did, not the flag itself:

> When I launch it from the command line, even without --keep-temp-files, it works.  If I double-click it to launch it (in its Applications folder), then it fails.

That makes sense to me: it’s a very different runtime environment (in particular, your shell customizations are loaded, unlike when launching from Finder) and e.g. it could be that you can’t even successfully run some php scripts in the Finder-launched environment. Maybe that’s indeed what’s going on here.

Thanks for this link, I forgot all about it!

**Comment by umpirsky** [comment]
@vslavik Thanks, in which version of Poedit will https://github.com/vslavik/poedit/commit/849367f0fa6de350d4d6ee9fbb1ed67b18f045fa be included, just for the reference for people that step across this issue?

**Comment by vslavik** [comment]
> @vslavik Thanks, in which version of Poedit will vslavik/poedit@849367f be included, just for the reference for people that step across this issue?

The next one (1.8.7), but this rmdir thing is a red herring, this change only affects debugging use and misbehaving extractors (which yours isn’t).
